### PR TITLE
Support Amazon Kinesis Data Streams Service

### PR DESCRIPTION
**Pull request by GrahamCampbell**
NB The generator command was unable to generate `DescribeStream`:

```
In StructureShape.php line 29:
                                                          
  [InvalidArgumentException]                              
  The member "StreamDescription.Shards" does not exists.  
                                                          

Exception trace:
  at /data/src/CodeGenerator/src/Definition/StructureShape.php:29
 AsyncAws\CodeGenerator\Definition\StructureShape->getMember() at /data/src/CodeGenerator/src/Generator/PaginationGenerator.php:105
 AsyncAws\CodeGenerator\Generator\PaginationGenerator->generateOutputPagination() at /data/src/CodeGenerator/src/Generator/PaginationGenerator.php:74
 AsyncAws\CodeGenerator\Generator\PaginationGenerator->generate() at /data/src/CodeGenerator/src/Generator/OperationGenerator.php:122
 AsyncAws\CodeGenerator\Generator\OperationGenerator->generate() at /data/src/CodeGenerator/src/Command/GenerateCommand.php:269
 AsyncAws\CodeGenerator\Command\GenerateCommand->generateService() at /data/src/CodeGenerator/src/Command/GenerateCommand.php:147
 AsyncAws\CodeGenerator\Command\GenerateCommand->generateServicesSequential() at /data/src/CodeGenerator/src/Command/GenerateCommand.php:84
 AsyncAws\CodeGenerator\Command\GenerateCommand->execute() at /data/vendor/symfony/console/Command/Command.php:256
 Symfony\Component\Console\Command\Command->run() at /data/vendor/symfony/console/Application.php:971
 Symfony\Component\Console\Application->doRunCommand() at /data/vendor/symfony/console/Application.php:290
 Symfony\Component\Console\Application->doRun() at /data/vendor/symfony/console/Application.php:166
 Symfony\Component\Console\Application->run() at /data/src/CodeGenerator/generate.php:20
 require() at /data/generate:32
```

### CODE REVIEW

**Comment by jderusse** [comment]
> NB The generator command was unable to generate `DescribeStream`:


Indeed, this is the first service that have pagination over nested objects.
I'll take care of this tomorrow.

File: src/Service/Kinesis/src/Result/ListStreamsOutput.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            throw new InvalidArgument('missing client injected in paginated result');
+        }
+        if (!$this->input instanceof ListStreamsInput) {
+            throw new InvalidArgument('missing last request injected in paginated result');
+        }
+        $input = clone $this->input;
+        $page = $this;
+        while (true) {
+            if ($page->getHasMoreStreams()) {
+                $input->setExclusiveStartStreamName($page->getStreamNames[-1]());
```
Comment (jderusse) [inline src/Service/Kinesis/src/Result/ListStreamsOutput.php:53]: Will be handled by #1011

**Comment by jderusse** [comment]
You should now be able to rebase on top of master and add the operation `DescribeStream`

**Comment by GrahamCampbell** [comment]
I can do at the weekend. :)

**Comment by jderusse** [comment]
I've rebased and push -force your PR to fix tests (and a bug in pagination).

**Comment by GrahamCampbell** [comment]
Nice, thanks!

File: manifest.json [hunk 2]
```diff
@@ -142,6 +142,45 @@
...
+                "PutRecord",
+                "PutRecords",
+                "RegisterStreamConsumer",
+                "RemoveTagsFromStream",
+                "SplitShard",
+                "StartStreamEncryption",
+                "StopStreamEncryption",
+                "StreamExists",
+                "StreamNotExists",
+                "UpdateShardCount"
```
Comment (Nyholm) [inline manifest.json:181]: Do you really need all of these?

**Review by Nyholm — COMMENTED** [review]
Could we make the changed in the code generator in a different PR?

File: manifest.json (hunk 2 above)
Comment (GrahamCampbell) [reply manifest.json:181]: No. I'm not yet sure what I need.

Comment (GrahamCampbell) [reply manifest.json:181]: I doubt I need any of the methods like `IncreaseStreamRetentionPeriod` since that would be managed by cloudformation.

**Comment by theophileds** [comment]
Hello,

I was about to contribute to Kinesis support until I found out that someone already did!
Are there any blockers on this pull request?

**Comment by Nyholm** [comment]
There are two things: 
1) Separate the code generator changes into a new PR
2) Only generate the operations needed. I doubt we need all operations. But if you do, then that is okey =)

**Comment by GrahamCampbell** [comment]
I am still not really sure what options I need, but I'd be happy to go with a very minimal selection, then add in some additional ones if required. @jderusse knows the state of this PR the best at the moment. I'm not sure if he had to make any manual adjustments to the generated code?

**Comment by Nyholm** [comment]
Could you please rebase this PR?
I've separated the CodeGenerator stuff in #1040.

File: src/Service/Kinesis/src/Enum/EncryptionType.php [hunk 3]
```diff
@@ -3,10 +3,10 @@
 namespace AsyncAws\Kinesis\Enum;
 
 /**
- * The encryption type used. This value is one of the following:.
+ * The server-side encryption type used on the stream. This parameter can be one of the following values:.
```
Comment (GrahamCampbell) [inline src/Service/Kinesis/src/Enum/EncryptionType.php:6]: php-cs-fixer incorrectly changed this doc. it should actually be

```
The server-side encryption type used on the stream.

This parameter can be one of the following values:

 - `NONE`: Do not encrypt the records in the stream.
 - `KMS`: Use server-side encryption on the records in the stream using a customer-managed AWS KMS key.

Comment (GrahamCampbell) [reply src/Service/Kinesis/src/Enum/EncryptionType.php:6]: note in particular the lack of a period after the colon

Comment (GrahamCampbell) [reply src/Service/Kinesis/src/Enum/EncryptionType.php:6]: > php-cs-fixer incorrectly changed this doc.

well, more that we were miss-using the short description area of the phpdoc. it is only meant to span either one line or one sentence.

Comment (jderusse) [reply src/Service/Kinesis/src/Enum/EncryptionType.php:6]: You're right, but it's not related to this PR. 
The code is generated, then cs-fixed, We can fix this later

**Comment by jderusse** [comment]
Phpstan and psalm errors are not related to this PR and are addressed in https://github.com/async-aws/aws/pull/1041

This PR si RFR for my side.

**Review by Nyholm — APPROVED** [review]

**Comment by Nyholm** [comment]
Thank you.

**Pull request merged**